src/elfpy/markets.py

```diff
@@ -67,67 +95,117 @@ def __init__(
+            raise ValueError(f'ERROR: Unknown trade type "{agent_action["action_type"]}".')
         # update market state
         self.update_market(market_deltas)
-        # TODO: self.update_LP_pool(wallet_deltas["fees"])
-        return wallet_deltas
-
+        self.update_spot_price_and_rate()
+        print(f"{self.get_market_step_string()}")
+        # TODO: self.update_LP_pool(agent_deltas["fees"])
+        agent_action.agent.update_wallet(agent_deltas)  # update user state since market doesn't know about users
+        if self.verbose:
+            print(f" agent  Δs ={agent_deltas.display()}\n market Δs ={market_deltas.display()}")
+            print(f" post-trade {agent_action.agent.status_report()}") if self.verbose else None
+    
     def update_market(self, market_deltas):
         """
         Increments member variables to reflect current market conditions
         """
-        for key, value in market_deltas.items():
-            assert np.isfinite(value), f"markets.update_market: ERROR: market delta key {key} is not finite."
-        self.share_reserves += market_deltas["d_base_asset"]
-        self.bond_reserves += market_deltas["d_token_asset"]
-        self.cum_base_asset_slippage += market_deltas["d_base_asset_slippage"]
-        self.cum_token_asset_slippage += market_deltas["d_token_asset_slippage"]
-        self.cum_base_asset_fees += market_deltas["d_base_asset_fee"]
-        self.cum_token_asset_fees += market_deltas["d_token_asset_fee"]
-        self.base_asset_orders += market_deltas["d_base_asset_orders"]
-        self.token_asset_orders += market_deltas["d_token_asset_orders"]
-        self.base_asset_volume += market_deltas["d_base_asset_volume"]
-        self.token_asset_volume += market_deltas["d_token_asset_volume"]
+        for field in market_deltas.__dataclass_fields__:
+            value = getattr(market_deltas, field)
+            if value: # check that it's instantiated
+                if field == "d_liquidity_pool_history":
+                    assert isinstance(value, list), f"markets.update_market: Error:"\
+                    + f" {field} has value={value} should be a dict"
+                elif field in ["d_share_fee_history","d_token_fee_history"]:
+                    assert isinstance(value, dict), f"markets.update_market: Error:"\
+                    + f" {field} has value={value} should be a dict"
+                else:
+                    assert np.isfinite(value), f"markets.update_market: ERROR: market delta key {field} is not finite."
+        self.share_reserves             += market_deltas.d_base_asset/self.share_price
+        self.bond_reserves              += market_deltas.d_token_asset
+        self.share_buffer               += market_deltas.d_share_buffer
+        self.bond_buffer                += market_deltas.d_bond_buffer
+        self.liquidity_pool             += market_deltas.d_liquidity_pool
+        if market_deltas.d_liquidity_pool_history != []:
```

Comment:
This conditional is strange. Are you checking if it is not a list, or checking if it is not empty? Rewrite with a more explicit conditional. For example, if you want to check if it is a list use `isinstance` and to check if it is empty use `len`.